### PR TITLE
Feature: a new mixing method for magnetic calculations

### LINKED ISSUE #3225: Feature: mix charge density and magnetic density separately
### Background

Currently, ABACUS calculate residual using charge density and magnetic density together. Ref: Journal of the Physical Society of Japan 82 (2013) 114706. This modification is expected to greatly improve the convergence calculation problem of magnetic materials.

### Describe the solution you'd like

Implement a new mixing method for magnetic calculations

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Notice Possible Changes of Behavior (Reminder only for developers)

_No response_

### Notice any changes of core modules (Reminder only for developers)

_No response_

### Notice Possible Changes of Core Modules (Reminder only for developers)

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### PR DESCRIPTION
**Pull request by WHUweiqingzhou**
### Linked Issue
Fix #3225 

### What's changed?
1. Add Functions to Allocate/Destroy/Calculate magnetic density array
2. Change mix_rho
3. Change Kerker and inner product with new rho input
4. change default value of mixing
5. remove charge_mixing::autoset()
6. Add corresponding documentation
7. Fix charge_mixing_test
8. update some CI tests by setting `mixing_gg0=0.0`

Note: 
This PR will introduce some warnings in CI tests even with `nspin=1` since it changes the default value of `mixing_gg0` from 0.0 to 1.0, you can fix them by adding `mixing_gg0=0.0` in INPUT file.

Please see detail and heavy tests in documentation [A new mixing method for magnetic calculations](https://ucoyxk075n.feishu.cn/docx/FuGvdcyIUot0KgxWpGBcCs3Nn4b)

### CODE REVIEW

File: source/module_base/module_mixing/plain_mixing.h [hunk 1]
```diff
@@ -108,12 +109,36 @@ class Plain_Mixing : public Mixing
...
 #ifdef _OPENMP
 #pragma omp parallel for schedule(static, 4096 / sizeof(FPTYPE))
 #endif
-        for (int i = 0; i < length; ++i)
+            for (int i = 0; i < length; ++i)
+            {
+                data[i] = data_in[i] + this->mixing_beta * F_tmp[i];
+            }
+        }
+        else if (GlobalV::NSPIN == 2)
```
Comment (Qianruipku) [inline source/module_base/module_mixing/plain_mixing.h]: Mix_matrix also uses mixing class. In that case, maybe it is wrong.

Comment (WHUweiqingzhou) [reply source/module_base/module_mixing/plain_mixing.h]: How about adding two input array, one contains `alpha`, another one contains corresponding `index_start`?

Comment (Qianruipku) [reply source/module_base/module_mixing/plain_mixing.h]: You can use a different global varible temporarily. Maybe TWO_BETA ...
Only when NSPIN == 2 && pw, TWO_BETA is ON? 
In the future, I will remove the use of global variable and refactor push_data function to make it more general.

File: source/module_elecstate/module_charge/charge_mixing.h [hunk 2]
```diff
@@ -45,18 +46,22 @@ class Charge_Mixing
...
     void Kerker_screen_real(double* rho);
+    void Kerker_screen_real_test(double* rho);
 
     /**
      * @brief Inner product of two complex vectors
      *
      */
     double inner_product_recip(std::complex<double>* rho1, std::complex<double>* rho2);
+    double inner_product_recip_new1(std::complex<double>* rho1, std::complex<double>* rho2);
+    double inner_product_recip_new2(std::complex<double>* rho1, std::complex<double>* rho2);
```
Comment (Qianruipku) [inline source/module_elecstate/module_charge/charge_mixing.h:64]: These new functions should add Unit Tests.

File: source/module_base/module_mixing/plain_mixing.h (hunk 1 above)
Comment (WHUweiqingzhou) [reply source/module_base/module_mixing/plain_mixing.h]: Done.

File: source/module_elecstate/module_charge/charge_mixing.h (hunk 2 above)
Comment (WHUweiqingzhou) [reply source/module_elecstate/module_charge/charge_mixing.h:64]: Done.

**Review by Qianruipku — APPROVED** [review]

**Pull request merged**